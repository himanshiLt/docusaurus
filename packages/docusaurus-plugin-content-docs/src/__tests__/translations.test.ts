/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import type {LoadedContent, LoadedVersion} from '../types';
import {CURRENT_VERSION_NAME} from '../constants';
import {
  getLoadedContentTranslationFiles,
  translateLoadedContent,
} from '../translations';
import type {DocMetadata} from '@docusaurus/plugin-content-docs';
import {updateTranslationFileMessages} from '@docusaurus/utils';

function createSampleDoc(doc: Pick<DocMetadata, 'id'>): DocMetadata {
  return {
    editUrl: 'any',
    lastUpdatedAt: 0,
    lastUpdatedBy: 'any',
    next: undefined,
    previous: undefined,
    permalink: 'any',
    slug: 'any',
    source: 'any',
    unversionedId: 'any',
    version: 'any',
    title: `${doc.id} title`,
    sidebar_label: `${doc.id} title`,
    description: `${doc.id} description`,
    ...doc,
  };
}

function createSampleVersion(
  version: Pick<LoadedVersion, 'versionName'>,
): LoadedVersion {
  return {
    label: `${version.versionName} label`,
    path: '/docs/',
    mainDocId: '',
    routePriority: undefined,
    sidebarFilePath: 'any',
    isLast: true,
    contentPath: 'any',
    contentPathLocalized: 'any',
    docs: [
      createSampleDoc({id: 'doc1'}),
      createSampleDoc({id: 'doc2'}),
      createSampleDoc({id: 'doc3'}),
      createSampleDoc({id: 'doc4'}),
      createSampleDoc({id: 'doc5'}),
    ],
    sidebars: {
      docs: [
        {
          type: 'category',
          label: 'Getting started',
          collapsed: false,
          link: {
            type: 'generated-index',
            slug: '/category/getting-started-index-slug',
            permalink: '/docs/category/getting-started-index-slug',
            title: 'Getting started index title',
            description: 'Getting started index description',
          },
          items: [
            {
              type: 'doc',
              id: 'doc1',
            },
            {
              type: 'doc',
              id: 'doc2',
            },
            {
              type: 'link',
              label: 'Link label',
              href: 'https://facebook.com',
            },
            {
              type: 'ref',
              id: 'doc1',
            },
          ],
        },
        {
          type: 'doc',
          id: 'doc3',
        },
      ],
      otherSidebar: [
        {
          type: 'doc',
          id: 'doc4',
        },
        {
          type: 'doc',
          id: 'doc5',
        },
      ],
    },
    ...version,
  };
}

const SampleLoadedContent: LoadedContent = {
  loadedVersions: [
    createSampleVersion({
      versionName: CURRENT_VERSION_NAME,
    }),
    createSampleVersion({
      versionName: '2.0.0',
    }),
    createSampleVersion({
      versionName: '1.0.0',
    }),
  ],
};

function getSampleTranslationFiles() {
  return getLoadedContentTranslationFiles(SampleLoadedContent);
}
function getSampleTranslationFilesTranslated() {
  const translationFiles = getSampleTranslationFiles();
  return translationFiles.map((translationFile) =>
    updateTranslationFileMessages(
      translationFile,
      (message) => `${message} (translated)`,
    ),
  );
}

describe('getLoadedContentTranslationFiles', () => {
  it('returns translation files', async () => {
    expect(getSampleTranslationFiles()).toMatchSnapshot();
  });
});

describe('translateLoadedContent', () => {
  it('does not translate anything if translation files are untranslated', () => {
    const translationFiles = getSampleTranslationFiles();
    expect(
      translateLoadedContent(SampleLoadedContent, translationFiles),
    ).toEqual(SampleLoadedContent);
  });

  it('returns translated loaded content', () => {
    const translationFiles = getSampleTranslationFilesTranslated();
    expect(
      translateLoadedContent(SampleLoadedContent, translationFiles),
    ).toMatchSnapshot();
  });
});
