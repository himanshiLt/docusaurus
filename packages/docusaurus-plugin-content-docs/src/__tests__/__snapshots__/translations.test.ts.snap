// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getLoadedContentTranslationFiles returns translation files 1`] = `
[
  {
    "content": {
      "sidebar.docs.category.Getting started": {
        "description": "The label for category Getting started in sidebar docs",
        "message": "Getting started",
      },
      "sidebar.docs.category.Getting started.link.generated-index.description": {
        "description": "The generated-index page description for category Getting started in sidebar docs",
        "message": "Getting started index description",
      },
      "sidebar.docs.category.Getting started.link.generated-index.title": {
        "description": "The generated-index page title for category Getting started in sidebar docs",
        "message": "Getting started index title",
      },
      "sidebar.docs.link.Link label": {
        "description": "The label for link Link label in sidebar docs, linking to https://facebook.com",
        "message": "Link label",
      },
      "version.label": {
        "description": "The label for version current",
        "message": "current label",
      },
    },
    "path": "current",
  },
  {
    "content": {
      "sidebar.docs.category.Getting started": {
        "description": "The label for category Getting started in sidebar docs",
        "message": "Getting started",
      },
      "sidebar.docs.category.Getting started.link.generated-index.description": {
        "description": "The generated-index page description for category Getting started in sidebar docs",
        "message": "Getting started index description",
      },
      "sidebar.docs.category.Getting started.link.generated-index.title": {
        "description": "The generated-index page title for category Getting started in sidebar docs",
        "message": "Getting started index title",
      },
      "sidebar.docs.link.Link label": {
        "description": "The label for link Link label in sidebar docs, linking to https://facebook.com",
        "message": "Link label",
      },
      "version.label": {
        "description": "The label for version 2.0.0",
        "message": "2.0.0 label",
      },
    },
    "path": "version-2.0.0",
  },
  {
    "content": {
      "sidebar.docs.category.Getting started": {
        "description": "The label for category Getting started in sidebar docs",
        "message": "Getting started",
      },
      "sidebar.docs.category.Getting started.link.generated-index.description": {
        "description": "The generated-index page description for category Getting started in sidebar docs",
        "message": "Getting started index description",
      },
      "sidebar.docs.category.Getting started.link.generated-index.title": {
        "description": "The generated-index page title for category Getting started in sidebar docs",
        "message": "Getting started index title",
      },
      "sidebar.docs.link.Link label": {
        "description": "The label for link Link label in sidebar docs, linking to https://facebook.com",
        "message": "Link label",
      },
      "version.label": {
        "description": "The label for version 1.0.0",
        "message": "1.0.0 label",
      },
    },
    "path": "version-1.0.0",
  },
]
`;

exports[`translateLoadedContent returns translated loaded content 1`] = `
{
  "loadedVersions": [
    {
      "contentPath": "any",
      "contentPathLocalized": "any",
      "docs": [
        {
          "description": "doc1 description",
          "editUrl": "any",
          "id": "doc1",
          "lastUpdatedAt": 0,
          "lastUpdatedBy": "any",
          "next": undefined,
          "permalink": "any",
          "previous": undefined,
          "sidebar_label": "doc1 title",
          "slug": "any",
          "source": "any",
          "title": "doc1 title",
          "unversionedId": "any",
          "version": "any",
        },
        {
          "description": "doc2 description",
          "editUrl": "any",
          "id": "doc2",
          "lastUpdatedAt": 0,
          "lastUpdatedBy": "any",
          "next": undefined,
          "permalink": "any",
          "previous": undefined,
          "sidebar_label": "doc2 title",
          "slug": "any",
          "source": "any",
          "title": "doc2 title",
          "unversionedId": "any",
          "version": "any",
        },
        {
          "description": "doc3 description",
          "editUrl": "any",
          "id": "doc3",
          "lastUpdatedAt": 0,
          "lastUpdatedBy": "any",
          "next": undefined,
          "permalink": "any",
          "previous": undefined,
          "sidebar_label": "doc3 title",
          "slug": "any",
          "source": "any",
          "title": "doc3 title",
          "unversionedId": "any",
          "version": "any",
        },
        {
          "description": "doc4 description",
          "editUrl": "any",
          "id": "doc4",
          "lastUpdatedAt": 0,
          "lastUpdatedBy": "any",
          "next": undefined,
          "permalink": "any",
          "previous": undefined,
          "sidebar_label": "doc4 title",
          "slug": "any",
          "source": "any",
          "title": "doc4 title",
          "unversionedId": "any",
          "version": "any",
        },
        {
          "description": "doc5 description",
          "editUrl": "any",
          "id": "doc5",
          "lastUpdatedAt": 0,
          "lastUpdatedBy": "any",
          "next": undefined,
          "permalink": "any",
          "previous": undefined,
          "sidebar_label": "doc5 title",
          "slug": "any",
          "source": "any",
          "title": "doc5 title",
          "unversionedId": "any",
          "version": "any",
        },
      ],
      "isLast": true,
      "label": "current label (translated)",
      "mainDocId": "",
      "path": "/docs/",
      "routePriority": undefined,
      "sidebarFilePath": "any",
      "sidebars": {
        "docs": [
          {
            "collapsed": false,
            "items": [
              {
                "id": "doc1",
                "type": "doc",
              },
              {
                "id": "doc2",
                "type": "doc",
              },
              {
                "href": "https://facebook.com",
                "label": "Link label (translated)",
                "type": "link",
              },
              {
                "id": "doc1",
                "type": "ref",
              },
            ],
            "label": "Getting started (translated)",
            "link": {
              "description": "Getting started index description (translated)",
              "permalink": "/docs/category/getting-started-index-slug",
              "slug": "/category/getting-started-index-slug",
              "title": "Getting started index title (translated)",
              "type": "generated-index",
            },
            "type": "category",
          },
          {
            "id": "doc3",
            "type": "doc",
          },
        ],
        "otherSidebar": [
          {
            "id": "doc4",
            "type": "doc",
          },
          {
            "id": "doc5",
            "type": "doc",
          },
        ],
      },
      "versionName": "current",
    },
    {
      "contentPath": "any",
      "contentPathLocalized": "any",
      "docs": [
        {
          "description": "doc1 description",
          "editUrl": "any",
          "id": "doc1",
          "lastUpdatedAt": 0,
          "lastUpdatedBy": "any",
          "next": undefined,
          "permalink": "any",
          "previous": undefined,
          "sidebar_label": "doc1 title",
          "slug": "any",
          "source": "any",
          "title": "doc1 title",
          "unversionedId": "any",
          "version": "any",
        },
        {
          "description": "doc2 description",
          "editUrl": "any",
          "id": "doc2",
          "lastUpdatedAt": 0,
          "lastUpdatedBy": "any",
          "next": undefined,
          "permalink": "any",
          "previous": undefined,
          "sidebar_label": "doc2 title",
          "slug": "any",
          "source": "any",
          "title": "doc2 title",
          "unversionedId": "any",
          "version": "any",
        },
        {
          "description": "doc3 description",
          "editUrl": "any",
          "id": "doc3",
          "lastUpdatedAt": 0,
          "lastUpdatedBy": "any",
          "next": undefined,
          "permalink": "any",
          "previous": undefined,
          "sidebar_label": "doc3 title",
          "slug": "any",
          "source": "any",
          "title": "doc3 title",
          "unversionedId": "any",
          "version": "any",
        },
        {
          "description": "doc4 description",
          "editUrl": "any",
          "id": "doc4",
          "lastUpdatedAt": 0,
          "lastUpdatedBy": "any",
          "next": undefined,
          "permalink": "any",
          "previous": undefined,
          "sidebar_label": "doc4 title",
          "slug": "any",
          "source": "any",
          "title": "doc4 title",
          "unversionedId": "any",
          "version": "any",
        },
        {
          "description": "doc5 description",
          "editUrl": "any",
          "id": "doc5",
          "lastUpdatedAt": 0,
          "lastUpdatedBy": "any",
          "next": undefined,
          "permalink": "any",
          "previous": undefined,
          "sidebar_label": "doc5 title",
          "slug": "any",
          "source": "any",
          "title": "doc5 title",
          "unversionedId": "any",
          "version": "any",
        },
      ],
      "isLast": true,
      "label": "2.0.0 label (translated)",
      "mainDocId": "",
      "path": "/docs/",
      "routePriority": undefined,
      "sidebarFilePath": "any",
      "sidebars": {
        "docs": [
          {
            "collapsed": false,
            "items": [
              {
                "id": "doc1",
                "type": "doc",
              },
              {
                "id": "doc2",
                "type": "doc",
              },
              {
                "href": "https://facebook.com",
                "label": "Link label (translated)",
                "type": "link",
              },
              {
                "id": "doc1",
                "type": "ref",
              },
            ],
            "label": "Getting started (translated)",
            "link": {
              "description": "Getting started index description (translated)",
              "permalink": "/docs/category/getting-started-index-slug",
              "slug": "/category/getting-started-index-slug",
              "title": "Getting started index title (translated)",
              "type": "generated-index",
            },
            "type": "category",
          },
          {
            "id": "doc3",
            "type": "doc",
          },
        ],
        "otherSidebar": [
          {
            "id": "doc4",
            "type": "doc",
          },
          {
            "id": "doc5",
            "type": "doc",
          },
        ],
      },
      "versionName": "2.0.0",
    },
    {
      "contentPath": "any",
      "contentPathLocalized": "any",
      "docs": [
        {
          "description": "doc1 description",
          "editUrl": "any",
          "id": "doc1",
          "lastUpdatedAt": 0,
          "lastUpdatedBy": "any",
          "next": undefined,
          "permalink": "any",
          "previous": undefined,
          "sidebar_label": "doc1 title",
          "slug": "any",
          "source": "any",
          "title": "doc1 title",
          "unversionedId": "any",
          "version": "any",
        },
        {
          "description": "doc2 description",
          "editUrl": "any",
          "id": "doc2",
          "lastUpdatedAt": 0,
          "lastUpdatedBy": "any",
          "next": undefined,
          "permalink": "any",
          "previous": undefined,
          "sidebar_label": "doc2 title",
          "slug": "any",
          "source": "any",
          "title": "doc2 title",
          "unversionedId": "any",
          "version": "any",
        },
        {
          "description": "doc3 description",
          "editUrl": "any",
          "id": "doc3",
          "lastUpdatedAt": 0,
          "lastUpdatedBy": "any",
          "next": undefined,
          "permalink": "any",
          "previous": undefined,
          "sidebar_label": "doc3 title",
          "slug": "any",
          "source": "any",
          "title": "doc3 title",
          "unversionedId": "any",
          "version": "any",
        },
        {
          "description": "doc4 description",
          "editUrl": "any",
          "id": "doc4",
          "lastUpdatedAt": 0,
          "lastUpdatedBy": "any",
          "next": undefined,
          "permalink": "any",
          "previous": undefined,
          "sidebar_label": "doc4 title",
          "slug": "any",
          "source": "any",
          "title": "doc4 title",
          "unversionedId": "any",
          "version": "any",
        },
        {
          "description": "doc5 description",
          "editUrl": "any",
          "id": "doc5",
          "lastUpdatedAt": 0,
          "lastUpdatedBy": "any",
          "next": undefined,
          "permalink": "any",
          "previous": undefined,
          "sidebar_label": "doc5 title",
          "slug": "any",
          "source": "any",
          "title": "doc5 title",
          "unversionedId": "any",
          "version": "any",
        },
      ],
      "isLast": true,
      "label": "1.0.0 label (translated)",
      "mainDocId": "",
      "path": "/docs/",
      "routePriority": undefined,
      "sidebarFilePath": "any",
      "sidebars": {
        "docs": [
          {
            "collapsed": false,
            "items": [
              {
                "id": "doc1",
                "type": "doc",
              },
              {
                "id": "doc2",
                "type": "doc",
              },
              {
                "href": "https://facebook.com",
                "label": "Link label (translated)",
                "type": "link",
              },
              {
                "id": "doc1",
                "type": "ref",
              },
            ],
            "label": "Getting started (translated)",
            "link": {
              "description": "Getting started index description (translated)",
              "permalink": "/docs/category/getting-started-index-slug",
              "slug": "/category/getting-started-index-slug",
              "title": "Getting started index title (translated)",
              "type": "generated-index",
            },
            "type": "category",
          },
          {
            "id": "doc3",
            "type": "doc",
          },
        ],
        "otherSidebar": [
          {
            "id": "doc4",
            "type": "doc",
          },
          {
            "id": "doc5",
            "type": "doc",
          },
        ],
      },
      "versionName": "1.0.0",
    },
  ],
}
`;
